generate_spec_sources.py

```diff
@@ -0,0 +1,134 @@
+SPECFILE_IN = sys.argv[1]
+SPECFILE_OUT = sys.argv[2]
+
+BEGIN_SOURCES_MARKER = "# BEGIN AUTOGENERATED SOURCES"
+END_SOURCES_MARKER = "# END AUTOGENERATED SOURCES"
+
+BEGIN_SETUP_MARKER = "# BEGIN AUTOGENERATED SETUP"
+END_SETUP_MARKER = "# END AUTOGENERATED SETUP"
+
+BEGIN_SETUP_TMPL = "%setup -q "
+END_SETUP_TMPL = " -n %{getarchivedir 0}"
+
+
+class Replacement:
+    """Class describing a replacement of lines between markers in a file"""
+
+    def __init__(self, begin_marker, end_marker, generate):
+        self.begin_marker = begin_marker
+        self.end_marker = end_marker
+        self.generate = generate
+        self.throwaway = False
+        self.marker_seen = False
+
+    def process_line(self, line):
+        # did we change the input line? If not, we are merely returning it.
+        changed = False
+        of = io.StringIO()
+        if not self.throwaway:
+            of.write(line)
+            if line.strip() == self.begin_marker:
+                if self.marker_seen:
+                    sys.exit("Duplicate replacement section")
+                changed = True
+                self.generate(of)
+                self.throwaway = True
+                self.marker_seen = True
+            elif line.strip() == self.end_marker:
+                sys.exit("End marker '{}' without begin marker".format(self.end_marker))
+        else:
+            if line.strip() == self.end_marker:
+                self.throwaway = False
+                of.write(line)
+            elif line.strip() == self.begin_marker:
+                sys.exit("Duplicate begin marker: {}".format(self.begin_marker))
+            else:
+                # we discarded the input line
+                changed = True
+
+        return (changed, of.getvalue())
+
+    def inside(self):
+        """Are we currently inside the replacement block?"""
+        return self.throwaway
+
+    def after(self):
+        """Have we been inside the replacement block?"""
+        return self.marker_seen and not self.throwaway
+
+
+with open(COLLECTION_RELEASE, "r") as cf:
+    roles = yaml.safe_load(cf)
+    for rolename, roleinfo in roles.items():
+        print(roleinfo)
+    rolesbysourcenum = {
+        roleinfo["sourcenum"]: rolename for rolename, roleinfo in roles.items()
+    }
+
+
+def generate_sources(output):
+    k = list(rolesbysourcenum.keys())
+    k.sort()
+
+    for i in k:
+        r = rolesbysourcenum[i]
+        org = roles[rolesbysourcenum[i]].get("org")
+        repo = roles[rolesbysourcenum[i]].get("repo")
+        ref = roles[rolesbysourcenum[i]]["ref"]
+        if org:
+            # If we support other hosting service than Github, the format
+            # of collection_release.yml will have to change a bit.
+            output.write(f"%global forgeorg{i} https://github.com/{org}\n")
+        if repo:
+            output.write(f"%global repo{i} {repo}\n")
+        output.write(
+            f"""%global rolename{i} {r}
+%deftag {i} {ref}
```

Comment:
This assumes `ref` from `collection_release.yml` is a tag which is a semantic version - what if it is a commit hash?

Reply:
Indeed - I have thought that the file is maintained by release_collection.py, which keeps track of original role versions, and thus automatically ensures that the refs are tags. But now I see that there is `--use-commit-hash-(role)`, which breaks this rule, although it is not in use currently.
Since release_collection.py knows whether it is using a commit hash or a tag as a ref, I would prefer to change the format of `collection_release.yml` a bit and let release_collection.py record this information, instead of forgetting it on write and then having to guess it here.

Reply:
Thank you for the explanation how it would commit hashes be used. I implemented the support for them.
That said, when I try to use a commit hash in collection_release.yml, it seems to break the  "Publish fedora.linux_system_roles collection to Ansible Galaxy" GitHub action. It aborts with this:
```
Exception: No new-version given and cannot auto-update version.  Please examine the changes to the collection, determine what the new semantic version will be, and re-run this script with the --new-version argument
  Error: Process completed with exit code 1.
```
https://github.com/pcahyna/auto-maintenance/actions/runs/3878083249/jobs/6613862726
The action is heavily modified by me, but I don't think that the error is due to this.
So, I suspect that this "rare event" of using a commit hash could break the automation if you ever have to resort to it.